bpf/lib/nodeport.h

```diff
@@ -119,6 +130,175 @@ static __always_inline int nodeport_nat_ipv6_rev(struct __sk_buff *skb,
 	return NODEPORT_DO_NAT_IPV6(addr, NAT_DIR_INGRESS);
 }
 
+# ifdef ENABLE_DSR
+
+static __always_inline int set_dsr_ext6(struct __sk_buff *skb,
+					struct ipv6hdr *ip6,
+					union v6addr *svc_addr, __be32 svc_port)
+{
+	struct dsr_opt_v6 opt = {};
+
+	opt.nexthdr = ip6->nexthdr;
+	ip6->nexthdr = NEXTHDR_DEST;
+	ip6->payload_len = bpf_htons(bpf_ntohs(ip6->payload_len) + 24);
+
+	opt.len = DSR_IPV6_EXT_LEN;
+	opt.opt_type = DSR_IPV6_OPT_TYPE;
+	opt.opt_len = DSR_IPV6_OPT_LEN;
+	ipv6_addr_copy(&opt.addr, svc_addr);
+	opt.port = svc_port;
+
+	if (skb_adjust_room(skb, sizeof(opt), BPF_ADJ_ROOM_NET, 0))
+		return DROP_INVALID;
+
+	if (skb_store_bytes(skb, ETH_HLEN + sizeof(*ip6), &opt, sizeof(opt), 0) < 0)
+		return DROP_INVALID;
+
+	return 0;
+}
+
+static __always_inline int find_dsr_v6(struct __sk_buff *skb, __u8 nexthdr,
```

Comment:
Maybe it makes sense to extend `ipv6_hdrlen` by adding a callback function which could be used to find the DSR extension option?

Reply:
Maybe if the callback is also a __always_inline it could work with clang so that it inlines everything, haven't tried.